File: ship_collision.py
import numpy as np
import time
from functools import wraps

def timer_decorator(func):
    @wraps(func)
    def wrapper(*args, **kwargs):
        start = time.perf_counter()  # High-precision timer
        result = func(*args, **kwargs)
        end = time.perf_counter()
        print(f"Function '{func.__name__}' executed in: {(end - start):.4f} seconds")
        return result
    return wrapper

# This does assume the ship is a dot rather than a box
def cpa_distance(pos1, vel1, pos2, vel2):
    rel_pos = pos2 - pos1
    rel_vel = vel2 - vel1
    rel_vel_norm2 = np.dot(rel_vel, rel_vel)
    
    if rel_vel_norm2 == 0:
        return np.linalg.norm(rel_pos) # No relative motion: constant distance
    
    t = -np.dot(rel_pos, rel_vel) / rel_vel_norm2
    if t < 0:  
        return np.linalg.norm(rel_pos) # Closest point in the past: use current distance
    
    closest_rel = rel_pos + t * rel_vel
    return np.linalg.norm(closest_rel)

def estimate_crossing_risk(num_trials, num_other_ships, area_size=10.0, max_speed=2.0, threshold_dist=0.2):
    count_risks = 0
    main_pos = np.array([0.0, 0.0])  
    main_vel = np.array([1.0, 0.0])
    
    for _ in range(num_trials):
        has_risk = False
        for _ in range(num_other_ships):
            # Random position for other ship
            other_pos = np.random.uniform(-area_size, area_size, 2)
            # Random direction and speed
            angle = np.random.uniform(0, 2 * np.pi)
            speed = np.random.uniform(0, max_speed)
            other_vel = speed * np.array([np.cos(angle), np.sin(angle)])

            dist = cpa_distance(main_pos, main_vel, other_pos, other_vel)
            if dist < threshold_dist:
                has_risk = True
                break  # At least one crossing detected
        if has_risk:
            count_risks += 1
    return count_risks / num_trials

@timer_decorator
def main():
    num_trials = 50000
    num_other_ships = 15
    area_size = 25.0  # Area is a 20x20 square centered at origin
    max_speed = 2.0  # Other ships up to twice the main ship's speed
    threshold_dist = 0.2  # Safety threshold (e.g., in km or nautical miles)
    prob = (estimate_crossing_risk(num_trials, num_other_ships, area_size, max_speed, threshold_dist) * 100)
    rounded_prob = float(f"{prob:.2f}")
    print(f"\nEstimated probability of collision: {rounded_prob}%")  

if __name__ == "__main__":
    main()

File: ship_collision.cpp
#include <iostream>
#include <array>
#include <random>
#include <cmath>

std::array<float, 2> randomUniformPosition(float areaSize) {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_real_distribution<float> dist(-areaSize, areaSize);

    std::array<float, 2> result;
    for (float& x : result) {
        x = dist(gen);
    }
    return result;
}

float randomUniform(float first, float second) {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_real_distribution<float> dist(first, second);

    return dist(gen);
}

float cpaDistance(std::array<float, 2> pos1, std::array<float, 2> vel1, std::array<float, 2> pos2, std::array<float, 2> vel2) {
    std::array<float, 2> relPos;
    relPos[0] = pos2[0] - pos1[0];
    relPos[1] = pos2[1] - pos1[1];

    std::array<float, 2> relVel;
    relVel[0] = vel2[0] - vel1[0];
    relVel[1] = vel2[1] - vel1[1];

    float relVelNorm2 = relVel[0] * relVel[0] + relVel[1] * relVel[1];  // Dot Product
    
    if (relVelNorm2 == 0) {
        return std::sqrt(relPos[0] * relPos[0] + relPos[1] * relPos[1]);
    }

    float t = -(relPos[0] * relVel[0] + relPos[1] * relVel[1]) / relVelNorm2;

    if (t < 0) {
        return std::sqrt(relPos[0] * relPos[0] + relPos[1] * relPos[1]);
    }

    std::array<float, 2> closestRel;
    closestRel[0] = relPos[0] + t * relPos[0];
    closestRel[1] = relPos[1] + t * relPos[1];

    return std::sqrt(closestRel[0] * closestRel[0] + closestRel[1] * closestRel[1]); 
}


float estimateCrossingRisk(float numTrials, int numOtherShips, int area, int maxSpeed, float thresholdDist) {
    int countCollisions = 0;
    std::array<float, 2> mainPos = {0.0, 0.0};
    std::array<float, 2> mainVel = {1.0, 0.0};

    for (int i = 0; i < numTrials; i++) {
        bool isCollision = false;

        for (int j = 0; j < numOtherShips; j++) {
            std::array<float, 2> otherPos = randomUniformPosition(area);
            float angle = randomUniform(0, 2 * M_PI);
            float speed = randomUniform(0, maxSpeed);
            
            std::array<float, 2> otherVel = {speed * std::cos(angle), speed * std::sin(angle)};
            // Function call to CPA here TODO
            float dist = cpaDistance(mainPos, mainVel, otherPos, otherVel);
            std::cout << "CPA = " << dist << '\n';
            if (dist < thresholdDist) {
                isCollision = true;
                break;
            }
        }
        if (isCollision) {
            countCollisions++;
        }
    }
    float rate = countCollisions / numTrials;
    return rate;
}


int main() {
    int num_trials = 50000;
    int num_other_ships = 15;
    int area_size = 25;
    int max_speed = 2;
    float threshold_dist = 0.5;
    float prob = (estimateCrossingRisk(num_trials, num_other_ships, area_size, max_speed, threshold_dist) * 100);
    std::cout << "\n Estimated probability of collision: " << prob << "%" << std::endl;
    
    return 0;
}